typescript/sdk/src/token/EvmERC20WarpRouteReader.ts

```diff
@@ -44,34 +69,63 @@ export class EvmERC20WarpRouteReader {
    */
   async deriveWarpRouteConfig(
     address: Address,
-  ): Promise<DerivedERC20WarpRouteConfig> {
+  ): Promise<DerivedTokenRouterConfig> {
     const fetchedBaseMetadata = await this.fetchBaseMetadata(address);
     const fetchedTokenMetadata = await this.fetchTokenMetadata(
       fetchedBaseMetadata.token,
     );
 
-    const results: DerivedERC20WarpRouteConfig = {
+    // Derive the config type
+    const type = await this.deriveTokenType(address);
+
+    const derivedTokenRouter: DerivedTokenRouterConfig = {
       ...fetchedBaseMetadata,
       ...fetchedTokenMetadata,
+      type,
+    };
+
+    return derivedTokenRouter;
+  }
+
+  /**
+   * Derives the token type for a given Warp Route address using specific methods
+   *
+   * @param address - The Warp Route address to derive the token type for.
+   * @returns The derived token type, which can be one of: collateralVault, collateral, native, or synthetic.
+   */
+  async deriveTokenType(address: Address): Promise<DerivedTokenType> {
+    const contractTypes: Record<
+      Exclude<DerivedTokenType, 'native'>, // native is excluded because it's the default return type
+      { factory: any; method: string }
+    > = {
+      collateral: {
+        factory: HypERC20Collateral__factory,
+        method: 'wrappedToken',
+      },
+      collateralVault: {
+        factory: HypERC20CollateralVaultDeposit__factory,
+        method: 'vault',
+      },
+      synthetic: {
+        factory: HypERC20__factory,
+        method: 'decimals',
+      },
     };
```

Comment:
honestly a bit confused by this
is this a static constant that can be pulled out of the class?

Reply:
All this function does is, given a contract `address`, try to derive it's TokenType by calling an associated on chain `method` in the list. 

For example, if it can call `wrappedToken()`, then it is a `collateral`. If it can call `decimals()`, then it is a `synthetic`. I wanted to keep this list in here so it's clear that we only support deriving these 3 types.

Reply:
open to suggestions btw :)